src/comm/TcpAcceptor.cc

```diff
@@ -440,6 +408,34 @@ Comm::TcpAcceptor::oldAccept(Comm::ConnectionPointer &details)
     }
 #endif
 
+    details->nfConnmark = Ip::Qos::getNfConnmark(details, Ip::Qos::dirAccepted);
+
+    if (Config.client_ip_max_connections >= 0) {
+        if (clientdbEstablished(details->remote, 0) > Config.client_ip_max_connections) {
```

Comment:
This call will create a client_db record even though we do not plan to accept this connection into Squid. The new zero-counter record will eventually be garbage-collected, but its creation just wastes resources when they are already tight. And when then limit is not reached, we are calling clientdbEstablished() at least twice for the same connection, which is confusing -- the connection is established/accepted only once.

I suggest either adjusting clientdbEstablished() so that it does not create a client_db record when delta is zero OR, better, adding a read-only clientdbCount() function with similar functionality that should be used _instead_ of the current hack of calling clientdbEstablished() with zero delta. The supposed-to-be-read-only ACL code will benefit from this adjustment as well.

This suggested improvement is outside this PR scope, and, naturally, I do not insist on it. On the other hand, if implementing it in this PR is easier/cheaper, I certainly would not object to enlarging this PR scope a bit. Your call.

Reply:
That one I think should be in a followup PR.

Reply:
Sounds good to me.